### ISSUE TITLE
Crash in XSS filter

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/Pale-Moon#385]
We're running into an issue with the XSS filter when using the browser's "back" button.

https://forum.palemoon.org/viewtopic.php?f=3&t=11476

The issue is that people using the back button on certain forum software
(Lithium-based) get app crashes. Attaching a debugger, I found the following
stack:

```
> xul.dll!nsIDocument::GetRootElement() Line 2166   C++
    xul.dll!nsXSSUtils::DecodeHTMLEntities(const nsAString_internal & aString,
nsAString_internal & result, nsIDocument * doc) Line 288    C++
    xul.dll!nsXSSUtils::UnescapeLoop(const nsAString_internal & aString,
nsAString_internal & result, nsIDocument * doc) Line 265    C++
    xul.dll!nsXSSFilter::PermitsJSUrl(const nsAString_internal & aURI) Line
319 C++
    xul.dll!nsJSThunk::EvaluateScript(nsIChannel * aChannel, PopupControlState
aPopupState, unsigned int aExecutionPolicy, nsPIDOMWindow *
aOriginalInnerWindow) Line 212  C++
    xul.dll!nsJSChannel::EvaluateScript() Line 748  C++
    xul.dll!nsRunnableMethodImpl<void (__thiscall
nsSocketTransportService::*)(void),1>::Run() Line 351   C++
    xul.dll!nsThread::ProcessNextEvent(bool mayWait, bool * result) Line 632    C++
    xul.dll!NS_ProcessNextEvent(nsIThread * thread, bool mayWait) Line 238  C++
```

The relevant code is:

``` C++
nsresult
nsXSSUtils::DecodeHTMLEntities(const nsAString& aString, nsAString& result,
                               nsIDocument* doc)
{
  // 2 bugs so far, both in svg reftest. one seems to be a random
  // failure, while the second seems to be triggered by the call to
  // setinnerhtml. specifically, an assert condition is violated.

  LOG_XSS("DecodeHTMLEntities");
  result = aString;

  Element *rootElement = doc->GetRootElement(); <<< **CRASH**
  if (rootElement) {
    LOG_XSS("Has root element.");
    if (rootElement->GetNameSpaceID() == kNameSpaceID_SVG) {
      LOG_XSS("SVG Doc, bailing out");
      return NS_OK; // SVG doesn't support setting a title
    }
  } else {
    LOG_XSS("No root element");
  }
```

The crashpoint is a little odd, but it's probably because GetRootElement()
is a virtual function in nsIDocument.h - it's actually calling:

``` C++
  virtual Element* GetNameSpaceElement()
  {
    return GetRootElement();  <<< **CRASH**
  }
```

**Comment by wolfbeast** [comment]
Pinging Riccardo since this is his code -- we probably need a nullcheck on `doc`, but unsure what should be returned in case of a bailout.

**Comment by wolfbeast** [comment]
There already is a null check in the calling function which makes this a rather strange error.

**Comment by wolfbeast** [comment]
Tagging @BruceBerry here for reference. Whenever you have time, don't let this distract you from your thesis :)

It seems the problem is that GetRootElement() crashes because mCachedRootElement is defined, and pointing to *something*, but that something is an invalid struct for this context, probably because the cached element isn't actually an element but a document fragment instead. A cached document fragment won't always have a root element.

We need some sort of sanity check here, or making sure we never use a cached version.

**Comment by wolfbeast** [comment]
I've tried working around this by making sure to clear the cached reference to the root element before calling GetRootElement() but that didn't fix the crash; it also only occurs after paging back and forth a couple of times (pulling data from the disk cache), and the debugger isn't very helpful because it halts at the virtual call (can't get a proper stack trace from it).

Is this a race condition?

**Comment by wolfbeast** [comment]
Also, why exactly is the decodeHTMLElements inside the do-while loop in nsXSSUtils::UnescapeLoop? That might spin too fast... 

Don't we need to do that only once per doc/request?

**Comment by wolfbeast** [comment]
It seems the situation occurs when there actually is no content to check (aString is typically `'';`) - a workaround for this would be to test aString's length and bail if it's too short. Preliminary tests show that checking for a length < 4 solves the crashes on these Lithium-based fora.

**Comment by wolfbeast** [comment]
Looks like we'll have to make the filter less sensitive by short-circuiting this particular function as it seems to cause more crashes, e.g. https://forum.palemoon.org/viewtopic.php?f=15&t=11676&p=84909#p84909

**Comment by wolfbeast** [comment]
The workaround works, and we won't have the XSS filter in Tycho unless/until it's made compatible. Putting this on hold.
